### ISSUE TITLE
v1.1.0 Discussion - New Frontend & Chat Playback

### ISSUE THREAD

**Issue by Zibbp** [Zibbp/ganymede#63]
## About

To update, simply pull down the latest containers `docker compose pull` and bring them up `docker compose up -d`.

[v1.1.0](https://github.com/Zibbp/ganymede/releases/tag/v1.1.0) is the largest Ganymede update to date. It includes many new features such as:
- New Frontend (React/NextJS)
- New Chat Playback mode (similar to a Twitch VOD)
- Watching channels for new videos

## Chat Playback
A new chat playback mode as been added with the new frontend. It is similar to watching a VOD on Twitch. If you prefer to use the rendered video chat, a setting to enable that can be found in the "profile" page.

https://user-images.githubusercontent.com/21207065/203616455-604a7a21-98eb-40fe-9e12-621b7daaa8a8.mp4

#### How it works
When navigating to watch a VOD Ganymede will first check if there are embedded emotes in the VOD's chat file. If so, those emotes will be exclusively used. If no emotes are not found, Ganymede will fetch global and channel emotes from Twitch, BTTV, FFZ, and 7TV. This same process occurs with badges.

When playback starts the frontend will request 10 second segments of the chat. While doing this, the API will cache the chat json to improve response times of this request. Caching is being used as some of the chat files can grow to hundreds of megabytes depending on the number of embedded emotes. Caching the chat significantly speeds up the chat requests for playback.

When a skip (seek) in time is detected, the chat is cleared and different chat request is called before the typical 10 second segment api call. This api call fetches 30 comments before the destination seek time to populate the empty chat panel.

#### Performance
Currently, performance (memory usage) is not fantastic. I need to work on optimizing the memory usage with the chat playback routes. Depending on the size of the `chat.json` file (some are upwards of 250mb), a temporary usage of 1gb-2gb of ram (only for the stupid large 250+mb chat files) can be used. This will decrease over a few minutes are the garbage collector works.

## New Frontend

The new frontend written in React using NextJS has been released. It can be found at the same repository as before https://github.com/Zibbp/ganymede-frontend

![image](https://user-images.githubusercontent.com/21207065/203608906-3f2337ef-a4a7-491a-82f9-88e0a299c821.png)

#### Environment Variable Changes
The new frontend includes some new env vars, and some changes to the existing ones.
**New**
Show a "login via sso" button on the login page
```
SHOW_SSO_LOGIN_BUTTON=true/false
```
Force SSO auth when clicking the "login" button on the navbar. This takes the user directly to the SSO provider.
```
FORCE_SSO_AUTH=true/false
```
**Updated**
The `NUXT_PUBLIC_API_URL` and `NUXT_PUBLIC_CDN_URL` env vars of the old frontend were renamed to `API_URL` and `CDN_URL`. Do note that the old env vars will still work.
```
API_URL=http://localhost:4000
CDN_URL=http://localhost:4001
```

# Issues / Suggestions

Have any issues or suggestions with the new features or new frontend? Feel free to post them below or create a new issue!

**Comment by soleny894** [comment]
Hi,

First, thanks for that update it's great. I updated my containers to the last version but I have an issue with the API one. The container crash in loop because it can't use my timezone variable:
```
# This is Kubernetes env type
- name: TZ
  value: Europe/Amsterdam
```

Logs in the container:

```
Version    : 
Git Hash   : 40a559ebafb266dfb27ee4dde638a16e86e4cf6d
Build Time : 2022-11-23T17:47:46Z
{"level":"info","time":"2022-11-23T21:19:15Z","message":"config file found at /data/config.json, loading"}
{"level":"debug","time":"2022-11-23T21:19:15Z","message":"config file loaded: /data/config.json"}

   ____    __
  / __/___/ /  ___
 / _// __/ _ \/ _ \
/___/\__/_//_/\___/ v4.9.0
High performance, minimalist Go web framework
https://echo.labstack.com
____________________________________O/_______
                                    O\
⇨ http server started on [::]:4000
{"level":"info","time":"2022-11-23T21:19:16Z","message":"authenticated with twitch"}
{"level":"error","error":"unknown time zone Europe/Amsterdam","time":"2022-11-23T21:19:20Z","message":"failed to load location"}
{"level":"info","time":"2022-11-23T21:19:20Z","message":"setting up check watched channel videos schedule"}
panic: time: missing Location in call to Time.In

goroutine 88 [running]:
time.Time.In(...)
	/usr/local/go/src/time/time.go:1105
github.com/go-co-op/gocron.(*trueTime).Now(0x454107?, 0x0)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/timeHelper.go:16 +0xad
github.com/go-co-op/gocron.(*Scheduler).now(0xc0001f4c00)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/scheduler.go:1165 +0x37
github.com/go-co-op/gocron.(*Scheduler).scheduleNextRun(0xe14660?, 0xc0005f0000)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/scheduler.go:163 +0x38
github.com/go-co-op/gocron.(*Scheduler).runContinuous(0xc0001f4c00, 0xc0005f0000)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/scheduler.go:569 +0x27
github.com/go-co-op/gocron.(*Scheduler).runJobs(...)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/scheduler.go:91
github.com/go-co-op/gocron.(*Scheduler).start(0xc0001f4c00)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/scheduler.go:86 +0xa6
github.com/go-co-op/gocron.(*Scheduler).StartAsync(0xc0002acf50?)
	/go/pkg/mod/github.com/go-co-op/gocron@v1.15.1/scheduler.go:78 +0x2e
github.com/zibbp/ganymede/internal/scheduler.(*Service).StartWatchVideoScheduler(0x0?)
	/app/internal/scheduler/scheduler.go:57 +0x232
created by github.com/zibbp/ganymede/internal/transport/http.NewHandler
	/app/internal/transport/http/handler.go:88 +0x985
```

Removing the TZ env fix the issue.

**Comment by Zibbp** [comment]
Hi, I'm glad you're enjoying the update.

> I updated my containers to the last version but I have an issue with the API one. The container crash in loop because it can't use my timezone variable:

This is because I forgot to include the `tzdata` package in the API container. I just added that in this commit https://github.com/Zibbp/ganymede/commit/a0f6d79994fa8ec4745a781c9fae1f1f8b5ca24c.

I also made it so the container will not crash if an invalid timezone is supplied, it will default to UTC if it can't parse the provided timezone https://github.com/Zibbp/ganymede/commit/cf9449f0fafad648de907b5bfd2b7baa63bcffef.

I'm going to wait on a few more bug fixes before pushing out a release so you'll have to keep the TZ env var removed until I'm able to push a new release.

**Comment by Zibbp** [comment]
I just released [v1.1.1](https://github.com/Zibbp/ganymede/releases/tag/v1.1.1) which includes the `tzdata` package so you should be able to include your `TZ` environment variable now.

**Comment by soleny894** [comment]
Hi again,

Thanks for the update @Zibbp, the env seem to work correctly now. I have a new issue with VOD this time. I archived an old VOD and tried to play it but I got this message: `Application error: a client-side exception has occurred (see the browser console for more information).` and console is saying this:
```
TypeError: vod.views is undefined
    NextJS 29
        VodTitleBar
        Xh
        Wk
        Vk
        Jk
        Jk
        Ok
        Hk
        J
        R
        53
        __webpack_require__
        3840
        __webpack_require__
        4448
        __webpack_require__
        3935
        __webpack_require__
        745
        __webpack_require__
        4534
        __webpack_require__
        1783
        __webpack_require__
        <anonymous>
        O
        <anonymous>
        webpackJsonpCallback
        <anonymous>
framework-ffee79c6390da51e.js:9:70052
    NextJS 29
        Mi
        callback
        ih
        lk
        jk
        Qk
        Qk
        Hk
        J
        R
        (Async: EventHandlerNonNull)
    53
        __webpack_require__
        3840
        __webpack_require__
        4448
        __webpack_require__
        3935
        __webpack_require__
        745
        __webpack_require__
        4534
        __webpack_require__
        1783
        __webpack_require__
        <anonymous>
        O
        <anonymous>
        webpackJsonpCallback
        <anonymous>
```

VODs that are from the old version do works without any issue.

**Comment by Zibbp** [comment]
Strange, how old was the VOD that was archived? The "views" column defaults to `1` so I'm guessing it's something with the VOD response from the Twitch API overwriting the default `1`. Would you mind commenting the VOD ID so I can take a look?

**Comment by soleny894** [comment]
It was not an old VOD, I was wrong on this sorry. I was thinking about another VOD that I downloaded. The VOD that got the issue was new (stream ended recently) and when I checked the "Views" option in the VOD settings it was empty, the default "1" was not there. I can't replicate that issue now since the VOD got views on it. I took a random channel VOD that have 0 view (1663526296), archived it and I got that issue again, the "Views" is undefined.

**Comment by Zibbp** [comment]
I pushed a commit to the frontend that will show "0 views" when the views property is not returned when fetching the video. This should prevent the error from popping up. https://github.com/Zibbp/ganymede-frontend/commit/a72d21b109be1745f9bf9d9df85d8914a7229239.

**Comment by Zibbp** [comment]
Closing this issue. Feel free to re-open or create a new issue if needed.
